File: tools/convsym/src/input/AS_Listing_Experimental.cpp

/* ------------------------------------------------------------ *
 * ConvSym utility version 2.12									*
 * Input wrapper for the AS listing format						*
 * ------------------------------------------------------------	*/

#include <cstdint>
#include <string>
#include <map>

#include <IO.hpp>

#include "InputWrapper.hpp"


struct Input__AS_Listing_Experimental : public InputWrapper {

	Input__AS_Listing_Experimental() {}
	~Input__AS_Listing_Experimental() {}

	void parse(SymbolTable& symbolTable, const char *fileName, const char * opts) {
		const int sBufferSize = 1024;
		if (*opts) {
			IO::Log(IO::warning, "-inopt is not supported by this parser");
		}

		// Variables
		uint8_t sBuffer[ sBufferSize ];
		std::multimap<uint32_t, std::string> SymbolMap;
		IO::FileInput input = IO::FileInput( fileName, IO::text );
		if ( !input.good() ) { throw "Couldn't open input file"; }
		uint32_t lastSymbolOffset = -1;		// tracks symbols offsets to ignore sections where PC is reset (mainly Z80 stuff)

		// For every string in a listing file ...
		while ( input.readLine( sBuffer, sBufferSize ) >= 0 ) {

			// Known issues for the Sonic 2 disassembly:
			//	* Some macros somehow define labels that looks like global ones (notably, _MOVE and such)
			//	* Labels from injected Z80 code sometimes make it to the list ...

			uint8_t* ptr = sBuffer;		// WARNING: dereffed type should be UNSIGNED, as it's used for certain range-based optimizations

			// Check if this line has file idention number (pattern: "(d)", where d is a decimal digit)
			if ( *ptr == '(' ) {
				bool foundDigits = false;
				ptr++;

				// Cycle through as long as found characters are digits
				while ( (unsigned)(*ptr-'0')<10 ) { foundDigits=true; ptr++; }

				// If digit pattern doesn't end with ")" as expected, or no digits were found at all, stop
				if ( *ptr++ != ')' || !foundDigits ) continue;
			}

			// Ensure line has a proper number (pattern: "ddddd/", where d is any digit)
			{
				bool foundDigits = false;
				while ( *ptr == ' ' ) { ptr++; } // skip spaces, if present

				// Cycle through as long as found characters are digits
				while ( (unsigned)(*ptr-'0')<10 ) { foundDigits=true; ptr++; }

				// If digit pattern doesn't end with "/" as expected, or no digits were found at all, stop
				if ( *ptr++ != '/' || !foundDigits ) continue;
			}

			// Ensure line has a proper offset (pattern: "hhhh :", where h is a hexidecimal character)
			while ( *ptr == ' ' ) { ptr++; }						// skip spaces, if present
			uint8_t* const sOffset = ptr;							// remember location, where offset should presumably start
			{
				bool foundHexDigits = false;

				// Cycle though as longs as proper hexidecimal characters are fetched
				while (	(unsigned)(*ptr-'0')<10 || (unsigned)(*ptr-'A')<6 ) {
					foundHexDigits = true;
					ptr++;
				}

				if ( *ptr == 0x00 ) continue;							// break if end of line was reached
				*ptr++ = 0x00;											// separate string pointered by "offset" variable from the rest of the line
				while ( *ptr == ' ' ) { ptr++; }						// skip spaces ...

				// If offset pattern doesn't end with ":" as expected, or no hex characters were found at all, stop
				if ( *ptr++ != ':' || !foundHexDigits ) continue;
			}

			// Check if line matches a label definition ...
			if ( *ptr == ' ' && *(ptr+1) == '(' ) continue;
			while ( *ptr && (ptr-sBuffer < 40) ) { ptr++; }			// align to column 40 (where line starts)
			while ( *ptr==' ' || *ptr=='\t' ) { ptr++; }			// skip spaces or tabs, if present
			uint8_t* const label = ptr;								// remember location, where label presumably starts...
			
			// The first character should be a latin letter (A..Z or a..z)
			if ( (unsigned)(*ptr-'A')<26 || (unsigned)(*ptr-'a')<26 ) {

				// Other characters should be letters are digits or _ char
				while ( 											
					(unsigned)(*ptr-'A')<26 || (unsigned)(*ptr-'a')<26 || (unsigned)(*ptr-'0')<10 || *ptr=='_'
				) { ptr++; }

				// If the word is the only on the line and it ends with ":", treat it as a label
				if ( *ptr==':' ) {
					*ptr++ = 0x00;		// separate string pointered by "label" variable from the rest of the line

					// Convert offset to uint32_t
					uint32_t offset = 0;
					for ( uint8_t* c = sOffset; *c; c++ ) {
						offset = offset*0x10 + (((unsigned)(*c-'0')<10) ? (*c-'0') : (*c-('A'-10)));
					}

					// Add label to the symbols table, if:
					//	1) Its absolute offset is higher than the previous offset successfully added
					//	2) When base offset is subtracted and the mask is applied, the resulting offset is within allowed boundaries
					if ( (lastSymbolOffset == (uint32_t)-1) || (offset >= lastSymbolOffset) ) {
						// Check if this is a label after ds or rs macro ...
						while ( *ptr==' ' || *ptr=='\t' ) { ptr++; }			// skip spaces or tabs, if present      
						if ( *ptr == 'd' && *(ptr+1) == 's' && *(ptr+2)=='.' ) continue;
						if ( *ptr == 'r' && *(ptr+1) == 's' && *(ptr+2)=='.' ) continue;

						const bool inserted = symbolTable.add(offset, (const char*)label);
						if (inserted) {
							lastSymbolOffset = offset;
						}
					}
					else {
						IO::Log( IO::debug, "Symbol %s at offset %X ignored: its offset is less than the previous symbol successfully fetched", label, offset );
					}
				}
			}
		}
	}

};

File: tools/crates/convsym/src/input/as_listing_experimental.rs
/// Input parser for AS listing format (experimental).

use std::io::{BufRead, BufReader, Read};

use anyhow::Result;

use crate::input::InputParser;
use crate::symbol_table::SymbolTable;

pub struct AsListingExperimental;

fn is_hex_char(c: u8) -> bool {
    (c.wrapping_sub(b'0') < 10) || (c.wrapping_sub(b'A') < 6)
}

fn is_digit(c: u8) -> bool {
    c.wrapping_sub(b'0') < 10
}

fn decode_hex(s: &[u8]) -> u32 {
    let mut offset: u32 = 0;
    for &c in s {
        let digit = if c.wrapping_sub(b'0') < 10 {
            (c - b'0') as u32
        } else {
            (c - b'A' + 10) as u32
        };
        offset = offset * 0x10 + digit;
    }
    offset
}

impl InputParser for AsListingExperimental {
    fn parse(&self, symbol_table: &mut SymbolTable, file_name: &str, opts: &str) -> Result<()> {
        if !opts.is_empty() {
            eprintln!("Warning: -inopt is not supported by this parser");
        }
        let _ = opts;

        let reader: Box<dyn Read> = if file_name == "-" {
            Box::new(std::io::stdin())
        } else {
            Box::new(std::fs::File::open(file_name)?)
        };
        let reader = BufReader::new(reader);

        let mut last_symbol_offset: u32 = u32::MAX;

        for line_result in reader.lines() {
            let line = match line_result {
                Ok(l) => l,
                Err(_) => break,
            };
            let bytes = line.as_bytes();
            let len = bytes.len();
            let mut pos = 0;

            // Check for file indentation number: "(d)"
            if pos < len && bytes[pos] == b'(' {
                pos += 1;
                let mut found_digits = false;
                while pos < len && is_digit(bytes[pos]) {
                    found_digits = true;
                    pos += 1;
                }
                if !found_digits || pos >= len || bytes[pos] != b')' {
                    continue;
                }
                pos += 1;
            }

            // Ensure line has proper line number: "ddddd/"
            {
                while pos < len && bytes[pos] == b' ' {
                    pos += 1;
                }
                let mut found_digits = false;
                while pos < len && is_digit(bytes[pos]) {
                    found_digits = true;
                    pos += 1;
                }
                if !found_digits || pos >= len || bytes[pos] != b'/' {
                    continue;
                }
                pos += 1;
            }

            // Ensure line has proper offset: "hhhh :"
            while pos < len && bytes[pos] == b' ' {
                pos += 1;
            }
            let offset_start = pos;
            {
                let mut found_hex = false;
                while pos < len && is_hex_char(bytes[pos]) {
                    found_hex = true;
                    pos += 1;
                }
                if pos >= len || !found_hex {
                    continue;
                }
            }
            let offset_end = pos;
            pos += 1; // skip separator
            while pos < len && bytes[pos] == b' ' {
                pos += 1;
            }
            if pos >= len || bytes[pos] != b':' {
                continue;
            }
            pos += 1;

            // Check if line matches a label definition
            if pos < len && bytes[pos] == b' ' && pos + 1 < len && bytes[pos + 1] == b'(' {
                continue;
            }

            // Align to column 40
            while pos < len && (pos - 0) < 40 {
                pos += 1;
            }
            // In the original, it aligns to column 40 from the start of sBuffer
            // We approximate by skipping to a reasonable position
            let col40 = if 40 < len { 40 } else { pos };
            pos = std::cmp::max(pos, col40);

            // Skip spaces/tabs
            while pos < len && (bytes[pos] == b' ' || bytes[pos] == b'\t') {
                pos += 1;
            }
            let label_start = pos;

            // First character should be a latin letter
            if pos >= len
                || !((bytes[pos].wrapping_sub(b'A') < 26) || (bytes[pos].wrapping_sub(b'a') < 26))
            {
                continue;
            }
            pos += 1;

            // Other characters: letters, digits, or '_'
            while pos < len
                && ((bytes[pos].wrapping_sub(b'A') < 26)
                    || (bytes[pos].wrapping_sub(b'a') < 26)
                    || (bytes[pos].wrapping_sub(b'0') < 10)
                    || bytes[pos] == b'_')
            {
                pos += 1;
            }

            // Must end with ':'
            if pos >= len || bytes[pos] != b':' {
                continue;
            }
            let label_end = pos;
            pos += 1;

            let offset = decode_hex(&bytes[offset_start..offset_end]);
            let label = String::from_utf8_lossy(&bytes[label_start..label_end]).to_string();

            // Check if offset is monotonically increasing
            if last_symbol_offset == u32::MAX || offset >= last_symbol_offset {
                // Check for ds. or rs. after label
                while pos < len && (bytes[pos] == b' ' || bytes[pos] == b'\t') {
                    pos += 1;
                }
                if pos + 2 < len
                    && ((bytes[pos] == b'd' && bytes[pos + 1] == b's' && bytes[pos + 2] == b'.')
                        || (bytes[pos] == b'r'
                            && bytes[pos + 1] == b's'
                            && bytes[pos + 2] == b'.'))
                {
                    continue;
                }

                if symbol_table.add(offset, &label) {
                    last_symbol_offset = offset;
                }
            } else if symbol_table.debug {
                eprintln!(
                    "Symbol {} at offset {:X} ignored: offset less than previous",
                    label, offset
                );
            }
        }

        Ok(())
    }
}
